File: typescript/src/dryRun.ts
/**
 * Dry-run diff API (proposal #8).
 *
 * Mirrors `rulang/dry_run.py`. Engine.dryRun() returns a DryRunResult with a
 * structured trace of what matched, what executed, and per-action diffs.
 */

export type ActionKind = "set" | "compound" | "workflow" | "return";

export interface MatchedRule {
  readonly index: number;
  readonly ruleText: string;
  readonly conditionResult: boolean;
  readonly executed: boolean;
}

export interface ActionChange {
  readonly path: string;
  readonly before: unknown;
  readonly after: unknown;
}

export interface ExecutedAction {
  readonly ruleIndex: number;
  readonly actionKind: ActionKind;
  readonly changes: readonly ActionChange[];
  readonly returned?: unknown;
}

export interface DryRunResult {
  readonly matchedRules: readonly MatchedRule[];
  readonly executedActions: readonly ExecutedAction[];
  readonly diff: Record<string, [unknown, unknown]>;
  readonly finalEntity: unknown;
  readonly returned: unknown;
  readonly anyMatched: boolean;
}

// --- Entity diff --------------------------------------------------------

function sameValue(a: unknown, b: unknown): boolean {
  if (a === b) return true;
  if (a === null || b === null) return false;
  if (typeof a !== typeof b) return false;
  if (typeof a !== "object") return false;
  // Arrays
  if (Array.isArray(a) && Array.isArray(b)) {
    if (a.length !== b.length) return false;
    for (let i = 0; i < a.length; i++) {
      if (!sameValue(a[i], b[i])) return false;
    }
    return true;
  }
  if (Array.isArray(a) !== Array.isArray(b)) return false;
  // Plain objects
  const ak = Object.keys(a as object);
  const bk = Object.keys(b as object);
  if (ak.length !== bk.length) return false;
  for (const k of ak) {
    if (!(k in (b as Record<string, unknown>))) return false;
    if (!sameValue((a as Record<string, unknown>)[k], (b as Record<string, unknown>)[k])) return false;
  }
  return true;
}

export function diffEntities(
  before: unknown,
  after: unknown,
  path = "",
): Record<string, [unknown, unknown]> {
  if (sameValue(before, after)) return {};

  const typeMismatch =
    typeof before !== typeof after ||
    Array.isArray(before) !== Array.isArray(after) ||
    before === null ||
    after === null;
  if (typeMismatch) {
    return { [path || "<root>"]: [before, after] };
  }

  if (Array.isArray(before) && Array.isArray(after)) {
    if (before.length !== after.length) {
      return { [path || "<root>"]: [before, after] };
    }
    const changes: Record<string, [unknown, unknown]> = {};
    for (let i = 0; i < before.length; i++) {
      Object.assign(changes, diffEntities(before[i], after[i], `${path}[${i}]`));
    }
    return changes;
  }

  if (typeof before === "object" && before !== null && after !== null) {
    const b = before as Record<string, unknown>;
    const a = after as Record<string, unknown>;
    const keys = new Set([...Object.keys(b), ...Object.keys(a)]);
    const changes: Record<string, [unknown, unknown]> = {};
    for (const k of keys) {
      const subPath = path ? `${path}.${k}` : k;
      const bv = k in b ? b[k] : undefined;
      const av = k in a ? a[k] : undefined;
      if (!(k in b)) changes[subPath] = [null, av];
      else if (!(k in a)) changes[subPath] = [bv, null];
      else Object.assign(changes, diffEntities(bv, av, subPath));
    }
    return changes;
  }

  return { [path || "<root>"]: [before, after] };
}

export function mergeIntoAggregate(
  aggregate: Record<string, [unknown, unknown]>,
  path: string,
  before: unknown,
  after: unknown,
): void {
  if (path in aggregate) {
    const originalBefore = aggregate[path][0];
    if (sameValue(originalBefore, after)) {
      delete aggregate[path];
    } else {
      aggregate[path] = [originalBefore, after];
    }
  } else {
    aggregate[path] = [before, after];
  }
}

export function deepCopy<T>(value: T): T {
  if (value === null || typeof value !== "object") return value;
  if (Array.isArray(value)) return value.map((v) => deepCopy(v)) as unknown as T;
  if (value instanceof Date) return new Date(value.getTime()) as unknown as T;
  if (value instanceof Map) {
    const out = new Map();
    for (const [k, v] of value.entries()) out.set(deepCopy(k), deepCopy(v));
    return out as unknown as T;
  }
  if (value instanceof Set) {
    return new Set([...value].map((v) => deepCopy(v))) as unknown as T;
  }
  const out: Record<string, unknown> = {};
  for (const [k, v] of Object.entries(value as object)) out[k] = deepCopy(v);
  return out as T;
}

File: python/src/rulang/dry_run.py
"""
Dry-run diff API (proposal #8).

`RuleEngine.dry_run()` evaluates rules against an entity and returns a
structured trace instead of just the final value. Intended for previewing,
debugging, and regression-testing rule changes.

The returned `DryRunResult` exposes:

- `matched_rules`: every rule with its condition result and whether it
  executed. In first_match mode, rules that matched but lost to an earlier
  rule appear with `executed=False` so UIs can surface them.
- `executed_actions`: ordered list of actions that ran, with per-path
  before/after values.
- `diff`: aggregated path → (before, after) across the whole run.
- `final_entity`: post-run entity snapshot.
- `returned`: the value `evaluate()` would have returned.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any, Literal


@dataclass(frozen=True)
class MatchedRule:
    index: int
    rule_text: str
    condition_result: bool
    executed: bool


@dataclass(frozen=True)
class ActionChange:
    """A single path-level change produced by an action."""

    path: str
    before: Any
    after: Any


@dataclass(frozen=True)
class ExecutedAction:
    """One action's trace: which rule it came from and what it changed."""

    rule_index: int
    action_kind: Literal["set", "compound", "workflow", "return"]
    changes: tuple[ActionChange, ...]
    returned: Any = None  # set only when action_kind == "return"


@dataclass(frozen=True)
class DryRunResult:
    matched_rules: tuple[MatchedRule, ...]
    executed_actions: tuple[ExecutedAction, ...]
    diff: dict[str, tuple[Any, Any]]
    final_entity: Any
    returned: Any

    @property
    def any_matched(self) -> bool:
        return any(m.executed for m in self.matched_rules)


# --- Entity diffing -----------------------------------------------------


def _diff_entities(before: Any, after: Any, path: str = "") -> dict[str, tuple[Any, Any]]:
    """Return {path: (before_value, after_value)} for paths that differ."""
    if _same(before, after):
        return {}

    # If types differ or either is a leaf, record the whole node.
    if type(before) is not type(after):
        return {path or "<root>": (before, after)}

    if isinstance(before, dict):
        changes: dict[str, tuple[Any, Any]] = {}
        keys = set(before.keys()) | set(after.keys())
        for k in keys:
            sub_path = f"{path}.{k}" if path else str(k)
            if k not in before:
                changes[sub_path] = (None, after[k])
            elif k not in after:
                changes[sub_path] = (before[k], None)
            else:
                changes.update(_diff_entities(before[k], after[k], sub_path))
        return changes

    if isinstance(before, list):
        if len(before) != len(after):
            return {path or "<root>": (before, after)}
        changes = {}
        for i, (b, a) in enumerate(zip(before, after)):
            sub_path = f"{path}[{i}]"
            changes.update(_diff_entities(b, a, sub_path))
        return changes

    # Objects with __dict__
    if hasattr(before, "__dict__") and hasattr(after, "__dict__"):
        changes = {}
        keys = set(vars(before).keys()) | set(vars(after).keys())
        for k in keys:
            sub_path = f"{path}.{k}" if path else str(k)
            b = getattr(before, k, None)
            a = getattr(after, k, None)
            changes.update(_diff_entities(b, a, sub_path))
        return changes

    # Leaf scalars
    return {path or "<root>": (before, after)}


def _same(a: Any, b: Any) -> bool:
    try:
        return a == b
    except Exception:  # noqa: BLE001
        return a is b
